File: backend/src/main/java/com/primatoos/backend/model/DailyReport.java
package com.primatoos.backend.model;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.JoinTable;
import jakarta.persistence.ManyToMany;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.OneToMany;
import jakarta.persistence.OrderBy;
import jakarta.persistence.PrePersist;
import jakarta.persistence.Table;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@Entity
@Table(name = "daily_reports")
@Getter
@Setter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
@Builder
public class DailyReport {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "work_order_id", nullable = false)
    private WorkOrder workOrder;

    @Column(nullable = false)
    private LocalDate date;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "filled_by_worker_id", nullable = false)
    private Worker filledByWorker;

    @Builder.Default
    @ManyToMany(fetch = FetchType.LAZY)
    @JoinTable(
            name = "daily_report_team_present",
            joinColumns = @JoinColumn(name = "daily_report_id"),
            inverseJoinColumns = @JoinColumn(name = "worker_id")
    )
    private Set<Worker> teamPresent = new HashSet<>();

    @Column(name = "start_time")
    private LocalTime startTime;

    @Column(name = "end_time")
    private LocalTime endTime;

    @Column(name = "weather_condition")
    private String weatherCondition;

    @Column(name = "extra_services", columnDefinition = "TEXT")
    private String extraServicesExecuted;

    @Column(name = "problems_found", columnDefinition = "TEXT")
    private String problemsFound;

    @Column(name = "pending_issues", columnDefinition = "TEXT")
    private String pendingIssuesGenerated;

    @Column(name = "materials_used", columnDefinition = "TEXT")
    private String materialsUsed;

    @Column(name = "materials_missing", columnDefinition = "TEXT")
    private String materialsMissing;

    @Column(name = "next_day_forecast", columnDefinition = "TEXT")
    private String forecastForNextDay;

    @Column(columnDefinition = "TEXT")
    private String notes;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private DailyReportStatus status;

    @Builder.Default
    @OneToMany(mappedBy = "dailyReport", cascade = CascadeType.ALL, orphanRemoval = true)
    @OrderBy("id ASC")
    private List<DailyReportItem> items = new ArrayList<>();

    @Builder.Default
    @OneToMany(mappedBy = "dailyReport", fetch = FetchType.LAZY)
    @OrderBy("id ASC")
    private List<DailyReportPhoto> photos = new ArrayList<>();

    @Setter(AccessLevel.NONE)
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @PrePersist
    protected void onCreate() {
        this.createdAt = LocalDateTime.now();
    }
}

File: frontend/src/types/dailyReport.ts
import type { WorkerSummaryResponse } from './worker'
import type { WorkOrderSummaryResponse } from './workOrder'

export type DailyReportStatus = 'DRAFT' | 'FINALIZED'

export type DailyReportItemStatus =
  | 'EXECUTED'
  | 'PARTIALLY_EXECUTED'
  | 'NOT_EXECUTED'
  | 'NOT_APPLICABLE'

export interface DailyReportItemResponse {
  id: number
  activityDescription: string
  status: DailyReportItemStatus
  reason: string | null
  observation: string | null
  newExpectedDate: string | null
}

export interface DailyReportItemRequest {
  activityDescription: string
  status: DailyReportItemStatus
  reason?: string | null
  observation?: string | null
  newExpectedDate?: string | null
}

export interface DailyReportPhotoResponse {
  id: number
  url: string
  createdAt: string
}

export interface DailyReportResponse {
  id: number
  workOrder: WorkOrderSummaryResponse
  date: string
  filledByWorker: WorkerSummaryResponse
  teamPresent: WorkerSummaryResponse[]
  startTime: string | null
  endTime: string | null
  weatherCondition: string | null
  extraServicesExecuted: string | null
  problemsFound: string | null
  pendingIssuesGenerated: string | null
  materialsUsed: string | null
  materialsMissing: string | null
  forecastForNextDay: string | null
  notes: string | null
  status: DailyReportStatus
  items: DailyReportItemResponse[]
  photos: DailyReportPhotoResponse[]
  createdAt: string
}

export interface DailyReportSummaryResponse {
  id: number
  workOrder: WorkOrderSummaryResponse
  date: string
  filledByWorker: WorkerSummaryResponse
  status: DailyReportStatus
}

export interface DailyReportCreateRequest {
  workOrderId: number
  date: string
  teamPresentWorkerIds?: number[]
  startTime?: string | null
  endTime?: string | null
  weatherCondition?: string | null
  extraServicesExecuted?: string | null
  problemsFound?: string | null
  pendingIssuesGenerated?: string | null
  materialsUsed?: string | null
  materialsMissing?: string | null
  forecastForNextDay?: string | null
  notes?: string | null
  items?: DailyReportItemRequest[]
}

export interface DailyReportUpdateRequest {
  date: string
  teamPresentWorkerIds?: number[]
  startTime?: string | null
  endTime?: string | null
  weatherCondition?: string | null
  extraServicesExecuted?: string | null
  problemsFound?: string | null
  pendingIssuesGenerated?: string | null
  materialsUsed?: string | null
  materialsMissing?: string | null
  forecastForNextDay?: string | null
  notes?: string | null
  items?: DailyReportItemRequest[]
}
